### PR TITLE
Add macOS Unified Access Logging parser

### LINKED ISSUE #5760: Deprecation of ASL logging, and asl table
# Blueprint

This issue was opened by @groob originally on the osql fork:

> ASL logging is deprecating on macOS, which keeps the osquery builds from moving to a newer darwin SDK.
>
> Suggestion: move the asl table out of core into an extension. Move to the 10.14 SDK

ASL logging was deprecated a while back in favor of Unified Logging, but Apple never gave us an API for _reading_, only for _writing_. For reading, we were limited to invoking the `log` command-line process or the Console.app GUI. Trail of Bits created an osquery table for reading the Unified Log this way, as an extension https://github.com/trailofbits/osquery-extensions/tree/master/darwin_unified_log

I think the path forward should open soon, if we implement an osquery table for macOS Unified Logging that uses the `OSLogStore::getEntries()` API (new in the 10.15 SDK), As @keeleysam suggested:

> it should be replaced with something that calls 
> https://developer.apple.com/documentation/logging/oslogstore/3204125-getentries which is new in 10.15

Related: #2487 where the issue was acknowledged but resolved by suppressing the deprecation warnings and staying on the old SDK.

### PR DESCRIPTION
**Pull request by puffyCid**
This PR fixes and address the comments on the awesome PR done by @GarretReece and others at #6904 that adds unified log parsing.  
This PR only changes a few lines from the original implmentation done at #6904 to fix the issues mentioned:
* Segfault - Should be fixed, ran multiple times over 20 million logs, no issues
* Tests - Added tests to this PR
* Website docs - I can try to open a PR for the website documentation if this PR gets merged

On Slack it was mentioned it would likely be easier to open a new PR rather than trying to edit an existing one. 

The original PR summary for #6904 is copied below for clarity
```
This code implements a virtual table for darwin that uses the OSLog framework to read and return entries from the unified system log. Prior versions of macOS did not have an api to access this data, necessitating an extension that invoked the log utility. macOS 10.15 makes this no longer necessary.

At the moment I've tested these changes only on a 10.15 vm. I'd be grateful to the community for feedback regarding building on earlier versions of macOS.

Closes #5760
```

Let me know if additional changes or improvements need to be made

Closes: #6904
Closes: #5760

### CODE REVIEW

**Comment by puffyCid** [comment]
Also I just wanted to mention that i am also trying to implement a raw UAL parsing support to osquery at https://github.com/puffyCid/osquery/tree/unifiedlogs.  
The benefit of raw parsing is that it will support older macos versions and it is slightly more forensic focus (raw parsing vs api).
Its ~60% done and it still has lots of work to be done, but just wanted to mention it, if others want to monitor/contribute/comment.  Im slowly working/chipping away on it between other osquery features im working on.
References:
* [libyal doc](https://github.com/libyal/dtformats/blob/main/documentation/Apple%20Unified%20Logging%20and%20Activity%20Tracing%20formats.asciidoc)
* [existing example raw log parser](https://github.com/ydkhatri/UnifiedLogReader)

**Comment by directionless** [comment]
> Also I just wanted to mention that i am also trying to implement a raw UAL parsing support to osquery

To respond to the side note... I know there are advantages, but parsing feels more fraught than API usage. There are going to be security and sanitization questions, and file format durability questions.

**Comment by puffyCid** [comment]
> > Also I just wanted to mention that i am also trying to implement a raw UAL parsing support to osquery
> 
> To respond to the side note... I know there are advantages, but parsing feels more fraught than API usage. There are going to be security and sanitization questions, and file format durability questions.

yep very good points. Its mainly a side project/research. There is only one open source raw parser im aware of https://github.com/ydkhatri/UnifiedLogReader. 
(format docs for those curious: https://github.com/libyal/dtformats/blob/main/documentation/Apple%20Unified%20Logging%20and%20Activity%20Tracing%20formats.asciidoc)

The main reason a raw parser would be desired is mainly a forensic/investigation one. When doing forensic analysis on a system that is compromised from an analysis perspective the investigator can't really trust the OS (due to possible rootkits/malware).  And also for older macs where the api doesnt exist.
Ex: A hypothetical scenario in which a rootkit was installed on a mac and was able to manipulate the logging framework to hide log entries related to the rootkit/other malware would cause the api to not get the full data vs raw parsing the actual files would potentially be unaffected.  

Again, hypothetical, ive never seen that happen, but when doing forensic investigations ideally the data is parsed independently from the OS apis/tools (if possible).

But that being said, i think the work done by adding UAL parsing via API is awesome and will very useful for everyone!

**Comment by elyscape** [comment]
Any chance of this getting some movement? This would be a really useful feature to have.

**Comment by puffyCid** [comment]
@elyscape I'm currently waiting for feedback
Once the PR gets reviewed and feedback I can make changes and hopefully it will get merged

File: osquery/tables/system/darwin/unified_log.mm [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        predicateWithLeftExpression:keyExp
+                    rightExpression:valExp
+                           modifier:NSDirectPredicateModifier
+                               type:kSupportedOps.at(op)
+                            options:0];
+    [preds addObject:pred];
+  }
+}
+
+QueryData genUnifiedLog(QueryContext& queryContext) {
```
Comment (alessandrogario) [inline osquery/tables/system/darwin/unified_log.mm:100]: A mandatory `WHERE` clause should be implemented to limit the scope of the query and avoid watchdog issues with simple `SELECT * FROM table` statements

**Review by alessandrogario — CHANGES_REQUESTED** [review]

**Comment by FritzX6** [comment]
@mike-myers-tob, @directionless,

IIRC @puffyCid is no longer part of the osquery slack and I am not sure if they are continuing to contribute to the osquery project any longer, not sure how you think it best to proceed in terms of keeping this PR from being orphaned.

**Comment by directionless** [comment]
@FritzX6 This work has continued in #7598

**Pull request closed without merging**